src/driver/mod.rs

```diff
@@ -35,29 +35,29 @@ struct NsmMessage<'a> {
     /// User-provided data for the request
     pub request: IoSlice<'a>,
     /// Response data provided by the NSM pipeline
-    pub response: IoSliceMut<'a>,
+    pub response: IoSlice<'a>,
 }
 
-/// Encode an NSM `Request` value into a vector.  
-/// *Argument 1 (input)*: The NSM request.  
-/// *Returns*: The vector containing the CBOR encoding.
-fn nsm_encode_request_to_cbor(request: Request) -> Vec<u8> {
-    serde_cbor::to_vec(&request).unwrap()
+/// Encode an NSM `Request` value into a vector.
+/// *Argument 1 (input)*: The NSM request.
+/// *Argument 2 (input)*: The buffer that will have the CBOR written into it.
+fn nsm_encode_request_to_cbor(request: Request, buf: &mut Vec<u8>) {
+    ciborium::into_writer(&request, buf).expect("buf's Writer returned an unexpected error");
```

Comment:
An alternative to this `expect` _could_ be to swap around how we enforce `NSM_REQUEST_MAX_SIZE`.

The current flow looks like this:
1. Create a resizable buffer, initialized with `NSM_REQUEST_MAX_SIZE` capacity,  and provide a request of unchecked size to `nsm_encode_request_to_cbor`
2. Successfully CBOR encode the request
3. Check the length of CBOR, returning an error if it exceeds `NSM_REQUEST_MAX_SIZE`

This could be turned on its head so that `into_writer` is given a non-resizable `buf`. If we encounter a [`WriteZero`](https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.WriteZero). This would remove the need for the additional length check as that would be enforced when writing: a write attempt would fail when it runs out of space to write.